### DISCUSSION TITLE
Audio Pairing with Image/Keyboard Response in a Sequence

### DISCUSSION THREAD

**Discussion by mcvc968** [jspsych/jsPsych#1110] [Q&A]
Hi, 

Programming novice here, I apologise in advance if these questions have obvious answers. 

I am trying to run a serial reaction-time task where a participant completes an image/keyboard response and I would like to play trial-specific sound at the start of each trial such that the image presented in the trial becomes associated with a specific sound. 

I have tried to associate the sounds, images, and their corresponding correct responses thus:

```js
       /* Right Keys */
      var A_key = {stimulus: ["/Users/cristina/Documents/Motor Task/L_Middle.png", "Users/cristina/Documents/Motor Task/Music Notes/Piano.ff.A5-[AudioTrimmer.com].mp3"], data: {test_part: 'test', correct_response: 'a'},};

      var S_key = {stimulus: ["/Users/cristina/Documents/Motor Task/L_Index.png", "Users/cristina/Documents/Motor Task/Music Notes/Piano.ff.B0-[AudioTrimmer.com].mp3"], data: {test_part: 'test', correct_response: 's'},};

      /* Left Keys */
      var K_key = {stimulus: ["/Users/cristina/Documents/Motor Task/R_Index.png", "Users/cristina/Documents/Motor Task/Music Notes/Piano.ff.B0-[AudioTrimmer.com].mp3"], data: {test_part: 'test', correct_response: 'k'},};

      var L_key = {stimulus: ["/Users/cristina/Documents/Motor Task/R_Middle.png", "Users/cristina/Documents/Motor Task/Music Notes/Piano.ff.B0-[AudioTrimmer.com].mp3"], data: {test_part: 'test', correct_response: 'l'},};
```
I need the participants to complete these image/keyboard responses in a series of sequences. I have therefore created sequences out of the above variables thus:

```js
var sequence_A_stim = [A_key,S_key,A_key,L_key,S_key,K_key,L_key,A_key,K_key,S_key,L_key,K_key];
```

Finally, I have integrated them into an image/keyboard response trial thus:

```js
var seqA_test = {
    type: 'image-keyboard-response',
    stimulus: jsPsych.timelineVariable('sequence_A_stim'),
    choices: ['a','s','k','l'],
    data: jsPsych.timelineVariable('A_data'),
    on_finish: function(data){
      data.correct = data.key_press == jsPsych.pluginAPI.convertKeyCharacterToKeyCode(data.correct_response);
    }
  }
```

This results in this error : `Trial level node is missing the "type" parameter.` 

and this error: `The AudioContext was not allowed to start. It must be resumed (or created) after a user gesture on the page`

I don't understand the top error. I know in the previous forum it was mentioned that jsPsych doesn't cope well with arrays in the timeline but I am not sure how to fix this and keep the order of the sequence intact. 

I am also unsure how to play the audio at the same time as the image appears without user input. 

Any help would be appreciated. Apologies again for the messy code.

**Comment by jodeleeuw** [comment]
Hi @mcvc968 ,

Can you please post the part of your code where you construct the timeline that is passed to the `timeline` parameter of `jsPsych.init()`?

**Reply by mcvc968** [reply]
[study 1 session 1 copy.TXT](https://github.com/jspsych/jsPsych/files/5351068/study.1.session.1.copy.TXT)

Hi. I am very sorry, I am not sure which part of the code you mean so I am attaching the full script here.

**Reply by mcvc968** [reply]
I just realised how long the code is. Let me know if I can shorten the code so that the part you are looking for is easier to find or if you could clarify which part of the code you are looking for so I can be more helpful.

**Reply by becky-gilbert** [reply]
@mcvc968 I think the problem is with this section, where you create the motor task procedure:
```javascript
var motor_task = {
    timeline: [ test_psedorand , fixation],
    timeline_variables: [sequence_A_stim, sequence_B_stim],
    repetitions: 1 // BTW there was a typo in "repetitions", fixed here
}
```
The timeline needs to be an array of objects, whereas you've created an array that contains two arrays. Similarly, `timeline_variables` needs to be an array of objects.
You could try [flattening these arrays](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat), so that both your motor task `timeline` and `timeline_variables` are arrays of objects, e.g.
```javascript
var all_stim = [sequence_A_stim, sequence_B_stim].flat();
```

**Comment by becky-gilbert** [comment]
Just wanted to let you know that this error `The AudioContext was not allowed to start. It must be resumed (or created) after a user gesture on the page` is just a warning that is shown when using audio. It's letting you know that the browser won't automatically play audio files until the user makes a 'gesture' on the page, e.g. a key press or button click. So as long as the participant has to interact with the page somehow before the first audio is supposed to play, then you can just ignore this warning. Otherwise add a keyboard response or button response trial to the start of your main timeline.

**Reply by mcvc968** [reply]
Thank you, this was really helpful!
